### PR TITLE
Code Model: Don't crash when adding element, deleting its text and then adding it again

### LINKED ISSUE #8423: Code Model: Adding an element, deleting it from the editor and then adding it again crashes

### PR DESCRIPTION
**Pull request by DustinCampbell**
Fixes #8423

In Code Model, we weakly cache the code elements created by a FileCodeModel so long as they're alive. Elements are removed from the cache when they are deleted using Code Models APIs (e.g. `FileCodeModel.Remove(...)`) or during a clean up pass on idle. However, that means that it's possible to have collisions in the cache if a Code Model consumer does something so:

1. Add an element, say a field called 'placeholder'.
2. Replace the text of the element in the buffer.
3. Add another field called 'placeholder'.

The steps above will likely result in a crash if the cache doesn't get cleaned up after the text is changed.

This PR adds code to check the cache before a new element is added. If an element with the same node key is found in the cache and that element is "dead" (that is, it's node key can't be resolved), we remove the old element from the cache.

Tagging @dotnet/roslyn-ide

### CODE REVIEW

**Comment by jmarolf** [comment]
:+1:

File: src/VisualStudio/Core/Impl/CodeModel/InternalElements/AbstractCodeElement.cs [hunk 1]
```diff
@@ -56,9 +56,8 @@ protected ProjectId GetProjectId()
 
         internal bool IsValidNode()
         {
-            var node = LookupNode();
-
-            if (node == null)
+            SyntaxNode node;
+            if (!TryLookupNode(out node) || node == null)
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Impl/CodeModel/InternalElements/AbstractCodeElement.cs]: Shouldn't TryLookupNode return false if node is null.  THen you don't need the || check.

File: src/VisualStudio/Core/Impl/CodeModel/InternalElements/CodeAccessorFunction.cs [hunk 2]
```diff
@@ -47,19 +47,33 @@ private bool IsPropertyAccessor()
 
         internal override SyntaxNode LookupNode()
         {
+            SyntaxNode node;
+            if (!TryLookupNode(out node))
+            {
+                throw Exceptions.ThrowEFail();
+            }
+
+            return node;
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Impl/CodeModel/InternalElements/CodeAccessorFunction.cs]: Extract into helper and use in all locations you need.

**Comment by CyrusNajmabadi** [comment]
:+1:

**Comment by DustinCampbell** [comment]
@dotnet-bot retest prtest/win/dbg/unit32 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit32/3530/
// Retest reason:

**Comment by DustinCampbell** [comment]
@dotnet-bot retest prtest/win/tst/unit32 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_tst_unit32/152/
// Retest reason:

**Comment by DustinCampbell** [comment]
@dotnet-bot retest prtest/win/dbg/unit32 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit32/3535/
// Retest reason:

**Comment by DustinCampbell** [comment]
@dotnet-bot retest prtest/win/dbg/unit32 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit32/3556/
// Retest reason:

**Comment by DustinCampbell** [comment]
@dotnet-bot retest prtest/win/tst/unit32 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_tst_unit32/152/
// Retest reason:

**Comment by DustinCampbell** [comment]
@dotnet-bot retest prtest/win/dbg/unit32 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit32/3558/
// Retest reason:

**Comment by DustinCampbell** [comment]
@dotnet-bot retest prtest/win/tst/unit32 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_tst_unit32/152/
// Retest reason:

**Comment by DustinCampbell** [comment]
@dotnet-bot retest prtest/win/dbg/unit32 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit32/3561/
// Retest reason:

**Comment by Pilchie** [comment]
@davkean is this hitting the unit test issues there were in master?  What's the right thing to do here?  Close and re-open the PR because of #6356?

**Comment by DustinCampbell** [comment]
@Pilchie I've created separate bugs because each failure was on a different extension. I suspect it's just a race due to msbuild building in parallel, so it's like the same problem, but it's probably not #6356. I'm curious as to why "tst" jobs showed up here again since they seem to not be working.

**Comment by DustinCampbell** [comment]
@dotnet-bot retest prtest/win/dbg/unit32 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit32/3564/
// Retest reason:

**Pull request merged**